src/Hazelcast.Net/Clustering/ClusterMembers.cs

```diff
@@ -326,8 +333,24 @@ public async Task<MembersUpdatedEventArgs> SetMembersAsync(int version, ICollect
             // anymore, see note in GetMember)
             _loadBalancer.SetMembers(members.Select(x => x.Id));
 
-            // if this is the initial list of members, determine how to connect to members
-            if (previous.Count == 0)
+            // members provided through the members view event always provide their own internal address,
+            // and may (e.g. for Kubernetes clusters) also provide their public address - but in a Cloud
+            // setup for instance, members are not aware of their public address. so, we try to "fix"
+            // the members using the AddressProvider map, if any.
+            if (_clusterState.AddressProvider.HasMap)
+            {
+                foreach (var member in members.Where(x => !x.HasPublicAddress))
+                    member.PublicAddress = _clusterState.AddressProvider.Map(member.Address);
+            }
+
+            // and then we need to determine whether to connect to members through their internal address
+            // or their public address - this is the role of the ConnectAddressResolver - it can be forced
+            // through configuration options, or determined through rules:
+            // - if one member responds on its internal address, assume internal addresses are OK
+            // - if enough (sample size) members respond only on their public address, use public addresses
+            // for performance reasons (and this is what other clients e.g. Java do) we determine this
```

Comment:
Updated the text + `usePublicAddresses` does not have to be `true` or `false` it can be `null` which means: please figure things out. Which is precisely what we do here.